File: py/queue.py
from __future__ import annotations
from singly_linked_list import SinglyLinkedList
from typing import TypeVar

T = TypeVar('T')

class Queue:
    def __init__(self: Queue, firstElem : T) -> Queue:
        if firstElem:
            self.list = SinglyLinkedList()
            self.offer(firstElem)
        else:
            self.list = SinglyLinkedList()

    def size(self: Queue) -> int:
        return self.list.size

    def isEmpty(self: Queue) -> bool:
        return self.size() == 0

    def peek(self: Queue) -> T:
        if self.isEmpty():
            raise RuntimeError('Queue Empty')
        return self.list.peekFirst()

    def poll(self: Queue) -> T:
        if self.isEmpty():
            raise RuntimeError('Queue Empty')
        return self.list.removeFirst()

    def offer(self: Queue, elem: T) -> None:
        self.list.addLast(elem)




if __name__ == '__main__':
    queue = Queue(5)
    for i in range(5):
        queue.offer(i)

    for i in range(3):
        print(queue.poll())

File: cpp/queue.cpp
#include "singly_linked_list.cpp"

template <typename T>
class Queue{
  public:
    Queue() = default;
    Queue(T firstElem){
      offer(firstElem);
    }

    int size(){
      return list.getSize();
    }

    bool isEmpty(){
      return size() == 0;
    }

    T peek(){
      if (isEmpty()) throw std::runtime_error("Queue Empty");
      return list.peekFirst();
    }

    T poll(){
      if (isEmpty()) throw std::runtime_error("Queue Empty");
      return list.removeFirst();
    }

    void offer(T elem){
      list.addLast(elem);
    }

    std::string toString(){
      return list.toString();
    }

  private:
    SinglyLinkedList<T> list;
};

int main(){
  Queue<int> queue;
  for (int i {}; i < 5; i++){
      queue.offer(i);
  }
  for (int i {}; i < 3; i++){
      std::cout<<queue.poll()<<std::endl;
  }
  return 0;
}
